### PR TITLE
IceVision integration

### LINKED ISSUE #119: Additonal Backbones for Object Detection Task
Object Detection task currently only seems to support FasterRCNN it would be good to have additional backbones such as RetinaNet or YoloV5 .

### LINKED ISSUE #480: `metric` argument unused in object detection and seq2seq
## 🐛 Bug

The metric argument in the object detection and seq2seq tasks is currently unused, see #473

### LINKED ISSUE #487: [RFC] Areas in object detection data are not used
## 🐛 Bug

The `area` keys in `ObjectDetectionData` are not currently being used for anything, do we need them?

cc @hzitoun

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Fixes #487
 Fixes #480
 Fixes #119

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/core/integrations/icevision/transforms.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+def collate(samples: Sequence[Dict[str, Any]]) -> Dict[str, Sequence[Any]]:
+    return {key: [sample[key] for sample in samples] for key in samples[0]}
+
+
+@requires_extras("image")
+def default_transforms(image_size: Tuple[int, int]) -> Dict[str, Callable]:
+    """The default transforms for object detection: convert the image and targets to a tensor, collate the batch."""
+    return {
+        # "pre_tensor_transform": ApplyToKeys(
```
Comment (tchaton) [inline flash/core/integrations/icevision/transforms.py:33]: remove ? Same below ?

File: flash/core/integrations/icevision/model.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+    def finalize(self) -> Dict[str, float]:
+        logs = super().finalize()
+        return {
+            "Precision (IoU=0.50:0.95,area=all)": logs["AP (IoU=0.50:0.95) area=all"],
+            "Recall (IoU=0.50:0.95,area=all,maxDets=100)": logs["AR (IoU=0.50:0.95) area=all maxDets=100"],
+        }
+
+
+class IceVisionTask(Task):
```
Comment (tchaton) [inline flash/core/integrations/icevision/model.py:44]: Let's implement adapters instead :)

File: flash/image/detection/model.py [hunk 3]
```diff
@@ -76,132 +52,38 @@ class ObjectDetector(Task):
...
-        model_name,
-        num_classes,
-        backbone,
-        fpn,
-        pretrained,
-        pretrained_backbone,
-        trainable_backbone_layers,
-        anchor_generator,
-        **kwargs,
-    ):
```
Comment (tchaton) [inline flash/image/detection/model.py:130]: Could we get a torchvision adapter instead of deleting it. Some users might rely on this. This is a BC.

File: flash/image/instance_segmentation/data.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    ):
+        self.image_size = image_size
+
+        super().__init__(
+            train_transform=train_transform,
+            val_transform=val_transform,
+            test_transform=test_transform,
+            predict_transform=predict_transform,
+            data_sources={
+                "coco": IceVisionParserDataSource(parser=COCOMaskParser),
```
Comment (tchaton) [inline flash/image/instance_segmentation/data.py:51]: Neat !

File: flash/image/instance_segmentation/model.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        **kwargs: Any,
+    ):
+        self.save_hyperparameters()
+
+        super().__init__(
+            num_classes=num_classes,
+            backbone=backbone,
+            head=head,
+            pretrained=pretrained,
+            metrics=metrics or [SimpleCOCOMetric(COCOMetricType.mask)],
```
Comment (tchaton) [inline flash/image/instance_segmentation/model.py:78]: Not sure you need to wrap into a list.

File: flash_examples/graph_classification.py [hunk 6]
```diff
@@ -12,14 +12,13 @@
...
-from flash.core.utilities.imports import _TORCH_GEOMETRIC_AVAILABLE
+from flash.core.utilities.imports import example_requires
 from flash.graph.classification.data import GraphClassificationData
 from flash.graph.classification.model import GraphClassifier
 
-if _TORCH_GEOMETRIC_AVAILABLE:
-    from torch_geometric.datasets import TUDataset
-else:
-    raise ModuleNotFoundError("Please, pip install -e '.[graph]'")
+example_requires("graph")
```
Comment (tchaton) [inline flash_examples/graph_classification.py:19]: Neat !

File: flash/core/integrations/icevision/model.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        self.model_type, self.model, adapter, self.backbone = metadata["fn"](
+            backbone_config,
+            num_classes,
+            image_size=image_size,
+            **kwargs,
+        )
+        self.adapter = adapter(model=self.model, metrics=metrics)
+
+    @classmethod
+    def available_backbones(cls, head: str) -> List[str]:
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash/core/integrations/icevision/model.py]: Done :smiley:

File: flash/image/instance_segmentation/model.py (hunk 5 above)
Comment (ethanwharris) [reply flash/image/instance_segmentation/model.py]: This is being passed through to the icevision lightningmodule, which expects a list

File: flash/core/integrations/icevision/model.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        self,
+        num_classes: int,
+        backbone: str,
+        head: str,
+        pretrained: bool = True,
+        optimizer: Union[Type[torch.optim.Optimizer], torch.optim.Optimizer] = torch.optim.Adam,
+        optimizer_kwargs: Optional[Dict[str, Any]] = None,
+        scheduler: Optional[Union[Type[_LRScheduler], str, _LRScheduler]] = None,
+        scheduler_kwargs: Optional[Dict[str, Any]] = None,
+        metrics: Optional['IceVisionMetric'] = None,
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash/core/integrations/icevision/model.py]: The metric changes for each task, so it wouldn't make sense to have a default here

File: flash/core/integrations/icevision/model.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+                num_workers=num_workers,
+                pin_memory=pin_memory,
+                shuffle=shuffle,
+                drop_last=drop_last,
+                sampler=sampler,
+            )
+        return dataset
+
+    def training_step(self, batch, batch_idx) -> Any:
+        return self.adapter.training_step(batch, batch_idx)
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash/core/integrations/icevision/model.py]: Done

File: flash/image/detection/model.py (hunk 3 above)
Comment (ethanwharris) [reply flash/image/detection/model.py:135]: The old torchvision stuff wasn't really working. All the models from before are now available in the icevision integration

File: flash/core/integrations/icevision/transforms.py (hunk 1 above)
Comment (ethanwharris) [reply flash/core/integrations/icevision/transforms.py]: Done

File: .github/workflows/ci-testing.yml [hunk 10]
```diff
@@ -137,7 +137,7 @@ jobs:
       run: |
         sudo apt-get install libsndfile1
         pip install matplotlib
-        pip install '.[image]' --pre --upgrade
+        pip install '.[image]' --pre --upgrade torchaudio
```
Comment (Borda) [inline .github/workflows/ci-testing.yml:140]: shall we rather have audio related requirements?

File: docs/source/reference/instance_segmentation.rst [hunk 11]
```diff
@@ -0,0 +1,31 @@
+
+.. _instance_segmentation:
```
Comment (Borda) [inline docs/source/reference/instance_segmentation.rst:2]: we can have these labels automatically generated by docs

File: flash/core/adapter.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+from torch import nn
+from torch.utils.data import DataLoader, Sampler
+
+import flash
+from flash.core.data.auto_dataset import BaseAutoDataset
+from flash.core.model import DatasetProcessor, Task, Wrapper
+
+
+class Adapter(DatasetProcessor, Wrapper, nn.Module):
```
Comment (Borda) [inline flash/core/adapter.py:25]: can we have docstring with purpose/motivation of this class?

File: flash/core/integrations/icevision/transforms.py [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+        BBoxesRecordComponent,
+        ClassMapRecordComponent,
+        FilepathRecordComponent,
+        ImageRecordComponent,
+        InstancesLabelsRecordComponent,
+        KeyPointsRecordComponent,
+        MasksRecordComponent,
+        RecordIDRecordComponent,
+    )
+    from icevision.tfms import A
```
Comment (Borda) [inline flash/core/integrations/icevision/transforms.py:37]: what a cool class name :D

**Review by Borda — COMMENTED** [review]
would it be possible to have the refactoring in separate PR (for example moving backbones) so it is easier to see what are the real changes?

File: flash/image/detection/model.py [hunk 14]
```diff
@@ -74,144 +46,40 @@ class ObjectDetector(Task):
...
-                pretrained,
-                pretrained_backbone,
-                trainable_backbone_layers,
-                anchor_generator,
-                **kwargs,
-            )
-        else:
-            ValueError(f"{model} is not supported yet.")
+        metadata = self.heads.get(head, with_metadata=True)
+        adapter = metadata["metadata"]["adapter"].from_task(
```
Comment (tchaton) [inline flash/image/detection/model.py:67]: That's really powerful :)

File: flash/core/adapter.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+        pass
+
+    def validation_epoch_end(self, outputs) -> None:
+        pass
+
+    def test_epoch_end(self, outputs) -> None:
+        pass
+
+
+class AdapterTask(Task):
```
Comment (tchaton) [inline flash/core/adapter.py:56]: Could the AdapterTask use the mixins ?

File: tests/image/detection/test_model.py [hunk 16]
```diff
@@ -61,45 +77,44 @@ def test_init():
...
-    torch.jit.save(model, path)
-    model = torch.jit.load(path)
-
-    out = model([torch.rand(3, 32, 32)])
-
-    # torchvision RCNN always returns a (Losses, Detections) tuple in scripting
-    out = out[1]
-
-    assert {"boxes", "labels", "scores"} <= out[0].keys()
+# @pytest.mark.skipif(not _IMAGE_TESTING, reason="image libraries aren't installed.")
```
Comment (tchaton) [inline tests/image/detection/test_model.py:100]: Any reason why jit doesn't work. Maybe let's add a TODO ?

**Review by tchaton — APPROVED** [review]
Awesome work !

**Review by ananyahjha93 — APPROVED** [review]

**Comment by ananyahjha93** [comment]
LGTM!
